### PR TITLE
[RISCV] Fix more boundary cases in immediate selection for Zdinx load/store on RV32.

### PR DESCRIPTION
**Pull request by topperc**
In order to support -unaligned-scalar-mem properly, we need to be more careful with immediates of global variables. We need to guarantee that adding 4 in RISCVExpandingPseudos won't overflow simm12. Since we don't know what the simm12 is until link time, the only way to guarantee this is to make sure the base address is at least 8 byte aligned.

There were also several corner cases bugs in immediate folding where we would fold an immediate in the range [2044,2047] where adding 4 would overflow. These are not related to unaligned-scalar-mem.

I have no interest in Zdinx. I only wanted to remove the assert for unaligned scalar memory. Having an assert was a bad way to handle that since asserts aren't present in release builds.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -2705,7 +2735,7 @@ bool RISCVDAGToDAGISel::SelectAddrRegImm(SDValue Addr, SDValue &Base,
     // Handle immediates in the range [-4096,-2049] or [2048, 4094]. We can use
     // an ADDI for part of the offset and fold the rest into the load/store.
     // This mirrors the AddiPair PatFrag in RISCVInstrInfo.td.
-    if (isInt<12>(CVal / 2) && isInt<12>(CVal - CVal / 2)) {
+    if (CVal >= -4096 && CVal < (4094 - RV32ZdinxRange)) {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:2738]:
```suggestion
    if (CVal >= -4096 && CVal <= (4094 - RV32ZdinxRange)) {
```

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 2]
```diff
@@ -2649,20 +2654,44 @@ bool RISCVDAGToDAGISel::SelectAddrRegRegScale(SDValue Addr,
...
-    return true;
+    // If this is non RV32Zdinx we can always fold.
+    if (!IsRV32Zdinx) {
+      Base = Addr.getOperand(0);
+      Offset = Addr.getOperand(1);
+      return true;
+    }
+
+    // For RV32Zdinx we need to have more than 4 byte alignment so we can add 4
+    // to the offset when we expandin RISCVExpandPseudoInsts.
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:2673]:
```suggestion
    // to the offset when we expand in RISCVExpandPseudoInsts.
```

**Review by dtcxzyw — APPROVED** [review]
LG

**Comment by cumcum8197** [comment]
d982882b7ee5931350e3d3899d61e94e8ba90824

**Pull request merged**